### PR TITLE
AKS Script

### LINKED ISSUE #126: AKS Script: Create a script version of setup documentation
Create a bash script from IaC-AKS guide to setup AKS Envs for automation purpose. 

## Acceptance Criteria
- Script to automatically create the AKS environment
- Self-contained script (doesn't change env settings)
- Pass parameters as environment variable or argument
- Run smokers based on argument/env var

References: #105

### PR DESCRIPTION
**Pull request by kforeverisback**
# Type of PR

- [ ] Documentation changes
- [ ] Code changes
- [ ] Test changes
- [ ] CI-CD changes
- [ ] GitHub Template changes
- [X] Automation Script changes

## Purpose of PR
Automate the cluster creation process
For testers: Usage:
```bash
cd IaC/scripts
./create-cluster-env.bash --help # will print help
./create-cluster-env.bash -s azure-subsription-name -n basename123

```
`./create-cluster-env.bash --help` output:
```bash
    ./create-cluster-env.bash --ngsa-prefix basename123 [Optional Args/Flags]
    ./create-cluster-env.bash -s azure-subs -n basename123 [Optional Args/Flags]

Required args:
    -n | --ngsa-prefix BASE_NAME    This will be the NGSA prefix for all resources
                                    Do not include punctuation - only use a-z and 0-9
                                    must be at least 5 characters long
                                    must start with a-z (only lowercase)
Optional args:
    -s | --subscription AZ_SUB      Azure Subscription Name or ID
    -e | --env ENVIRONMENT          Environemnt Type. Default: dev (See README.md for other values)
    -d | --domain DOMAIN_NAME       Registered Domain Name. Default: nip.io (Requires --email)
    -m | --email EMAIL_DOMAIN       Required Email if a '--domain' is given
    -l | --location LOCATION        Location where the resources will be created. Default: westus2
    -k | --k8s-ver K8S_VERSION      Kubernetes version used. Default: 1.18.8
                                    Use 'az aks get-versions -l westus2 -o table' to get supported versions
    -c | --node-count NODE_COUNT    Cluster Node Count. Default: 3
    -r | --dns-rg DNS_RG            DNS Resource group name. Default: dns-rg
    -i | --cosmos-key COSMOS_KEY
    -u | --cosmos-url COSMOS_URL    In case users want to use their own CosmosDBBoth Key and URL are empty by default.
Optional Flag:
    -x | --set-k8s-context          Sets the kubernetes context for current user in /home/kushal/.kube/config
    -o | --smoke-it                 Set it to true or false to enable smokers deployment. Default: false
    -h | --help                     Show the usage
```
## Validation

- [X] Ran test and verified manually
- [ ] Update documentation or issue referenced above

## Issues Closed or Referenced

- Closes #126 
- References #105

### CODE REVIEW

**Comment by kforeverisback** [comment]
Did a `git reset` and tested again. It seems ngsa Helm deployment fails, because "zone" and "region" is `nil` in helm-config.example.  But if I set them manually, it works. Any thoughts?

**Comment by sivamu** [comment]
Looks like we would need to define sample values for the zone and region inside `helm-config.example.yaml` file. Helm should still be able to take the empty string as a value, not sure why it's getting nil. I'll look into why the default values are failing.

File: IaC/scripts/create-cluster-env.bash [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+    -d | --domain DOMAIN_NAME       Registered Domain Name. Default: nip.io (Requires --email)
+    -m | --email EMAIL_DOMAIN       Required Email if a '--domain' is given
+    -l | --location LOCATION        Location where the resources will be created. Default: westus2
+    -k | --k8s-ver K8S_VERSION      Kubernetes version used. Default: 1.18.8
+                                    Use 'az aks get-versions -l westus2 -o table' to get supported versions
+    -c | --node-count NODE_COUNT    Cluster Node Count. Default: 3
+    -r | --dns-rg DNS_RG            DNS Resource group name. Default: dns-rg
+    -i | --cosmos-key COSMOS_KEY    
+    -u | --cosmos-url COSMOS_URL    In case users want to use their own CosmosDBBoth Key and URL are empty by default. 
+Optional Flag:
```
Comment (sivamu) [inline IaC/scripts/create-cluster-env.bash:36]: Are the flags below optional as well? or the ones above should be renamed as required flags?

Comment (kforeverisback) [reply IaC/scripts/create-cluster-env.bash:36]: Yes All flags and arguments are optional except `-n | --ngsa-prefix`.

File: IaC/scripts/create-cluster-env.bash [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+envsubst < helm-config.example.yaml >| helm-config.yaml
+cd $REPO_ROOT/IaC/AKS/cluster/charts
+helm install ngsa-aks ngsa -f "./ngsa/helm-config.yaml" --namespace ngsa --set cert.issuer=letsencrypt-prod --set cert.enabled=${cert_enabled} --atomic
+
+# check the version endpoint
+# you may get a timeout error, if so, just retry
+http ${Ngsa_App_Endpoint}/version
+## Running validation
+# run the tests in a container
+docker run -it --rm retaildevcrew/webvalidate --server $Ngsa_Https_App_Endpoint --base-url https://raw.githubusercontent.com/retaildevcrews/ngsa/main/TestFiles/ --files baseline.json
```
[the remark this answers is not in the document]
Comment (sivamu) [reply IaC/scripts/create-cluster-env.bash:381]: I've updated the AKS readme for the latest usage, which is:
`docker run -it --rm retaildevcrew/loderunner:beta --server $Ngsa_Https_App_Endpoint --files baseline.json`

File: IaC/scripts/create-cluster-env.bash [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+    kubectl create namespace fluentbit
+    kubectl create secret generic fluentbit-secrets --namespace fluentbit --from-literal=WorkspaceId=${Ngsa_Log_Analytics_Wid} --from-literal=SharedKey=${Ngsa_Log_Analytics_SharedKey}
+
+    helm install fluentbit fluentbit --namespace fluentbit --atomic
+
+    # Verify the fluentbit pod is running
+    kubectl get pod --namespace fluentbit
+fi
+#echo -e "Removing temp dirs ${rmdirs[@]}"
+#for rd in ${rmdirs};do rm -r ${rd}; done
```
Comment (sivamu) [inline IaC/scripts/create-cluster-env.bash:405]: Still needed?
Also, GH linter error, need to add newline at end

File: IaC/scripts/create-cluster-env.bash [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+    echo -e "\nForwarding FluentBit log to Azure Log Analytics"
+    cd $REPO_ROOT/IaC/AKS/cluster/charts
+
+    kubectl create namespace fluentbit
+    kubectl create secret generic fluentbit-secrets --namespace fluentbit --from-literal=WorkspaceId=${Ngsa_Log_Analytics_Wid} --from-literal=SharedKey=${Ngsa_Log_Analytics_SharedKey}
+
+    helm install fluentbit fluentbit --namespace fluentbit --atomic
+
+    # Verify the fluentbit pod is running
+    kubectl get pod --namespace fluentbit
```
Comment (sivamu) [inline IaC/scripts/create-cluster-env.bash:402]: I don't think fluentbit should be dependent on the smoker install. Maybe separate it into it's own conditional block or always install fluentbit after the app setup?

File: IaC/scripts/create-cluster-env.bash [hunk 5]
```diff
@@ -0,0 +1,398 @@
+#!/bin/bash
+## TODO:
+##   Put descriptive echo's in Color
+##   Cleanup script [sem-done]
+## Qurey
+##   Zone and region needed to be selected for both smoker and ngsa
```
[the remark this answers is not in the document]
Comment (AAkindele) [reply IaC/scripts/create-cluster-env.bash:6]: depending on how much there is to change, you can probably make a new issue to update the references.

File: IaC/scripts/create-cluster-env.bash [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+
+# Sane prog env:
+set -o errexit -o pipefail -o noclobber # -o nounset
+aze() { echo "executing: az $@";az "$@"; }
+dockere() { echo "executing: docker $@";docker "$@"; }
```
[the remark this answers is not in the document]
Comment (kforeverisback) [reply IaC/scripts/create-cluster-env.bash:11]: They are not used. It's a one liner for debugging. I'll remove them.

File: IaC/scripts/create-cluster-env.bash (hunk 5 above)
[the remark this answers is not in the document]
Comment (kforeverisback) [reply IaC/scripts/create-cluster-env.bash:6]: There are three "smoke" words. It's an easy change.

File: IaC/scripts/create-cluster-env.bash (hunk 2 above)
[the remark this answers is not in the document]
Comment (kforeverisback) [reply IaC/scripts/create-cluster-env.bash:381]: I'll make similar changes in the script.

**Review by AAkindele — APPROVED** [review]
👍 i was able to run the script successfully.

**Pull request merged**